pkg/controller/syncset/syncset_controller.go

```diff
@@ -76,5 +83,97 @@ func (r *ReconcileSyncSet) Reconcile(request reconcile.Request) (reconcile.Resul
 		logger.WithField("elapsed", dur).Info("reconcile complete")
 	}()
 
+	var syncSet CommonSyncSet
+	if request.Namespace == "" {
+		syncSet = (*SelectorSyncSetAsCommon)(&hivev1.SelectorSyncSet{})
+	} else {
+		syncSet = (*SyncSetAsCommon)(&hivev1.SyncSet{})
+	}
+	err := r.Get(context.TODO(), request.NamespacedName, syncSet.AsRuntimeObject())
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return
+			return reconcile.Result{}, nil
+		}
+		logger.WithError(err).Error("error looking up syncset")
+		return reconcile.Result{}, err
+	}
+
+	// If the syncset is deleted, do not reconcile.
+	if syncSet.AsMetaObject().GetDeletionTimestamp() != nil {
+		logger.Debug("syncset is being deleted, nothing to do")
+		return reconcile.Result{}, nil
+	}
+
+	if syncSet.AsMetaObject().GetGeneration() == syncSet.GetStatus().ObservedGeneration {
+		logger.Debug("syncset generation has not changed, nothing to do")
+		return reconcile.Result{}, nil
+	}
+
+	var resourceErr error
+	resources := make([]hivev1.SyncResourceReference, 0, len(syncSet.GetSpec().Resources))
+	for i, res := range syncSet.GetSpec().Resources {
+		logger := logger.WithField("resource", i)
+		u := unstructured.Unstructured{}
+		switch err := yaml.Unmarshal(res.Raw, &u); {
+		case err != nil:
+			logger.WithError(err).Warn("could not unmarshal resource")
+			resourceErr = err
+		case u.GetName() == "":
+			logger.Warn("missing Name")
+			resourceErr = fmt.Errorf("missing Name")
+		default:
+			resources = append(resources,
+				hivev1.SyncResourceReference{
+					APIVersion: u.GetAPIVersion(),
+					Kind:       u.GetKind(),
+					Namespace:  u.GetNamespace(),
+					Name:       u.GetName(),
+				},
+			)
+		}
+		if resourceErr != nil {
+			setInvalidResourceCondition(syncSet, i, resourceErr)
+			break
+		}
+	}
+
+	if resourceErr == nil {
+		clearInvalidResourceCondition(syncSet)
+	}
+
+	syncSet.GetStatus().ObservedGeneration = syncSet.AsMetaObject().GetGeneration()
+	syncSet.GetStatus().Resources = resources
+
+	if err := r.Status().Update(context.Background(), syncSet.AsRuntimeObject()); err != nil {
+		logger.WithError(err).Log(controllerutils.LogLevel(err), "could not update syncset")
+		return reconcile.Result{}, err
+	}
+
 	return reconcile.Result{}, nil
 }
+
+func setInvalidResourceCondition(syncSet CommonSyncSet, resourceIndex int, resourceErr error) {
+	syncSet.GetStatus().Conditions = []hivev1.SyncSetCondition{{
+		Type:               hivev1.SyncSetInvalidResourceCondition,
+		Status:             corev1.ConditionTrue,
+		Reason:             "InvalidResource",
+		Message:            fmt.Sprintf("Resource %d is invalid: %v", resourceIndex, resourceErr),
+		LastProbeTime:      metav1.Now(),
+		LastTransitionTime: metav1.Now(),
+	}}
+}
+
+func clearInvalidResourceCondition(syncSet CommonSyncSet) {
+	if conds := syncSet.GetStatus().Conditions; len(conds) == 0 {
+		return
```

Comment:
The type of the condition is `InvalidResource`. We only want to add the condition if the status of the condition if `True`. Generally, conditions are not added if the condition is `False`. The `False` status is used because it is considered bad practice to remove a condition. So, if there an no invalid resources and have never been any invalid resources, then there should not be an `InvalidResource` condition on the syncset.

As for the name, the `setInvalidResourceCondition` function *sets* the `InvalidResource` condition, meaning that it will indicate that criteria are met for the condition. The `clearInvalidResourceCondition` function *clears* the `InvalidResource` condition, meaning that it will indicate that the criteria are not met for the condition.

Reply:
Okay. I guess for someone indoctrinated into the common idioms of "conditions" in k8s [1] this would be natural. To me it seems bizarre and convoluted.

[1] I'm not. I still have a devil of a time interpreting conditions where all the fields seem to contradict each other.